internal/ssh/ssh.go

```diff
@@ -0,0 +1,137 @@
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+// JUJU_SSH_DEFAULT_PORT is the default port we expect the juju controllers to respond on.
+const JUJU_SSH_DEFAULT_PORT = 2223
+
+// Resolver is the interface with the methods needed by the ssh jump server to route request.
+type Resolver interface {
+	// GetAddrFromModelUUID is the method to resolve the address of the controller to contact given the model UUID.
+	GetAddrFromModelUUID(ctx context.Context, user openfga.User, modelUUID string) (string, error)
+}
+
+// fowardMessage is the struct holding the information about the jump message received by the ssh client.
+type forwardMessage struct {
+	DestAddr string
+	DestPort uint32
+	SrcAddr  string
+	SrcPort  uint32
+}
+
+// Server is the custom struct to embed the gliderlabs.ssh server and a resolver.
+type Server struct {
+	*ssh.Server
+
+	resolver Resolver
+}
+
+// NewJumpSSHServer creates the jump server struct.
+func NewJumpSSHServer(ctx context.Context, port int, resolver Resolver) (Server, error) {
+	zapctx.Info(ctx, "NewSSHServer")
+	server := Server{
+		Server: &ssh.Server{
+			Addr: fmt.Sprintf(":%d", port),
+			ChannelHandlers: map[string]ssh.ChannelHandler{
+				"direct-tcpip": directTCPIPHandler(resolver),
+			},
+			PublicKeyHandler: func(ctx ssh.Context, key ssh.PublicKey) bool {
+				return true
+			},
+		},
+		resolver: resolver,
+	}
+
+	return server, nil
+}
+
+func directTCPIPHandler(resolver Resolver) func(srv *ssh.Server, conn *gossh.ServerConn, newChan gossh.NewChannel, ctx ssh.Context) {
+	return func(srv *ssh.Server, conn *gossh.ServerConn, newChan gossh.NewChannel, ctx ssh.Context) {
+		d := forwardMessage{}
+
+		k := newChan.ExtraData()
+
+		if err := gossh.Unmarshal(k, &d); err != nil {
+			rejectConnectionAndLogError(ctx, newChan, "Failed to parse channel data", err)
+			return
+		}
+		if d.DestPort == 0 {
+			d.DestPort = JUJU_SSH_DEFAULT_PORT
+		}
+		addr, err := resolver.GetAddrFromModelUUID(ctx, openfga.User{}, d.DestAddr)
+		if err != nil {
+			rejectConnectionAndLogError(ctx, newChan, "Failed to resolve address from model uuid", err)
+			return
+		}
+		dest := fmt.Sprintf("%s:%d", addr, d.DestPort)
+		// this is temporary. The way we dial to the controller will heavily change.
+		client, err := gossh.Dial("tcp", dest, &gossh.ClientConfig{
+			//nolint:gosec // this will be removed once we handle hostkeys
+			HostKeyCallback: gossh.InsecureIgnoreHostKey(),
+			Auth: []gossh.AuthMethod{
+				gossh.PasswordCallback(func() (secret string, err error) {
+					return "jwt", nil
+				}),
+			},
+		})
+		if err != nil {
+			rejectConnectionAndLogError(ctx, newChan, fmt.Sprintf("Failed to connect to %s: %v", dest, err), err)
+			return
+		}
+
+		dChan, reqs, err := client.OpenChannel("direct-tcpip", gossh.Marshal(d))
+		if err != nil {
+			rejectConnectionAndLogError(ctx, newChan, "Failed to open destination channel", err)
+			return
+		}
+
+		go gossh.DiscardRequests(reqs)
```

Comment:
a bit of explanation in the form of a godoc would be in order here and below.. why are we discarding certain requests..